### PR TITLE
simplify client type management

### PR DESCRIPTION
**Pull request by mkaczanowski**
The [ADR specifies](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R60) that a new Wasm light client type is "dynamically defined" during the light client binary upload.
Example CLI invocation:
```
$ ./simd tx ibc wasm-manager push-wasm "wasm/example-chain" binary.wasm
```

Our initial implementation used the client type as a namespace/alias, where every time the new binary is being uploaded under the same client type, we would version it. For example, the storage key `<client-type>/{latest, previous, next}` would point to selected versions of the binary.

While this looks nice from a user perspective, it raises a governance/ownership question. What prevents "other people" from uploading malicious binary under selected "client-type"? Who owns the "namespace/alias/client-type"?

This change simplifies the client type management:
1. all Wasm light clients are of `10-wasm` client type. This way we don't have to make any changes to [`IsAllowedClient`](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R60) method
2. every time Wasm blob is uploaded, the caller gets [`code-id` reference](https://github.com/cosmos/ibc-go/pull/163/files#diff-6e67f71bfd3b039c42b2c26c3f2d51430f78c348018096e99a7b4addbc042971R81) to uploaded code. This removes the governance requirement.

Implementation question (cc: @ethanfrey):
during the code upload, we compute `code_hash = sha256(wasm blob)`, to check if the code has been already uploaded. If it wasn't we store it under the `code_hash` key and call `code_id = WasmVM.Create`. Whenever a user calls the Wasm light client, he must provide the `code_id`.

So now, we have `code_hash` and `code_id` that identify the same binary. As it came out both IDs are exactly the same, meaning that CosmWasm also uses sha265 checksum.

Do you think it's a valid approach to assume that code_id and code_hash are always equal (do we control that)? If yes, then we could remove `code_hash` and use only the `code_id`.

### CODE REVIEW

File: docs/ibc/proto-docs.md [hunk 1]
```diff
@@ -3585,10 +3583,7 @@ WASM light client's Client state
 | ----- | ---- | ----- | ----------- |
 | `data` | [bytes](#bytes) |  |  |
 | `code_id` | [bytes](#bytes) |  |  |
-| `frozen` | [bool](#bool) |  |  |
```
Comment (mkaczanowski) [inline docs/ibc/proto-docs.md:3588]: this isn't required anymore since `IsFrozen` method has been replaced with `Status`

**Comment by ethanfrey** [comment]
> during the code upload, we compute code_hash = sha256(wasm blob), to check if the code has been already uploaded. If it wasn't we store it under the code_hash key and call code_id = WasmVM.Create. Whenever a user calls the Wasm light client, he must provide the code_id.
>
> So now, we have code_hash and code_id that identify the same binary. As it came out both IDs are exactly the same, meaning that CosmWasm also uses sha265 checksum.
>
> Do you think it's a valid approach to assume that code_id and code_hash are always equal (do we control that)? If yes, then we could remove code_hash and use only the code_id.

CosmWasm (both `wasmvm` and `cosmwasm-vm`) uses the `sha256(code)` as the unique identifier for the code. This ensures we can verify the authenticity and detect any corruption of files. We use the term [`checksum` to describe it in the API](https://github.com/CosmWasm/wasmvm/blob/main/src/cache.rs#L119). You can rely on this being stable.

In `wasmd`, we wanted simpler ids and to store extra metadata, so we made an `uint64` as a pimary key counter (1, 2, 3...) and call that `CodeID`. This will look up some metadata on the code used by `x/wasm` as well as the `checksum` used for all interactions with the Rust libraries. You can see where we link [CodeHash and CodeID in the code upload](https://github.com/CosmWasm/wasmd/blob/master/x/wasm/keeper/keeper.go#L175-L197).

You are welcome to expose the code hash directly (and trust the rust lib works on this) or make your own code_id as a lookup.

You can also avoid doing the `sha256` yourself, and just record the checksum returned when you `save_wasm` in the rust library. 

Hope this answers your questions

**Comment by mkaczanowski** [comment]
@ethanfrey that is a great answer, thanks a lot!

File: modules/core/28-wasm/keeper/keeper.go [hunk 2]
```diff
@@ -60,12 +60,10 @@ func NewKeeper(cdc codec.BinaryCodec, key sdk.StoreKey, vmConfig *VMConfig, vali
 	}
 }
 
-func (k Keeper) PushNewWASMCode(ctx sdk.Context, clientType string, code []byte) ([]byte, string, error) {
+func (k Keeper) PushNewWASMCode(ctx sdk.Context, code []byte) ([]byte, string, error) {
```
Comment (joe-bowman) [inline modules/core/28-wasm/keeper/keeper.go]: `s/WASM/Wasm/`

Comment (joe-bowman) [reply modules/core/28-wasm/keeper/keeper.go]: Throughout this file, in fact.

**Review by joe-bowman — APPROVED** [review]
there appear to be a couple of WASMs that have been missed, but otherwise, this feels like a great improvement. Thank you.

File: modules/core/28-wasm/keeper/keeper.go (hunk 2 above)
Comment (mkaczanowski) [reply modules/core/28-wasm/keeper/keeper.go]: yeah, I wanted to do it in separate commit (to not mix up things too much)

**Pull request merged**